[RNMobile] Sync VideoPress metadata when post is manually saved

SiobhyB opened the pull request:
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5666

## Related PRs

* `Gutenberg`: https://github.com/WordPress/gutenberg/pull/49781
* `Gutenberg Mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5663
* `iOS`: https://github.com/wordpress-mobile/WordPress-iOS/pull/20522
* `Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/18279

## Proposed changes:

* https://github.com/Automattic/jetpack/pull/30131/commits/43ba4149891c773d7f8be6685828138510d58e3a: The temporary "Save Settings" button (added in https://github.com/Automattic/jetpack/pull/29997) has been removed. 
* https://github.com/Automattic/jetpack/pull/30131/commits/85881662851ba106ac6969414bb135fe13497abf: A listener has been set up to listener for "save events" that come when a user manually saves a post in the native apps. Data will now only be synced from the app to the web when a save event has been detected.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

p9ugOq-3mV-p2

## Does this pull request change what data or activity we track or use?

No, it doesn't.

## Testing instructions:

* In the app, either add a new VideoPress block or navigate to an existing one.
* Tweak the block's settings. It's particularly important to change the settings in the **Details** and **Privacy and Rating** panels, as these are the settings that are not currently being synced as expected.
* Manually save your changes in the app. You can find different flows for manually saving on both Android and iOS below.
* Confirm that you see the expected logs in the console, showing that the settings have been synced.
* Navigate to the block on the web and verify that the block's settings have been correctly synced there.
* Go through a few tries of editing the block on the app then returning the web, and vice versa, to ensure syncing is working as expected.

### Android flows

<details>
 <summary>Flows for saving a draft ⤵️</summary>

* Tap the three dots to the upper right and then **Save**.
* Tap the back button.
* Tap the **Publish** button.
* The the three dots to the upper right to get to **Post Settings**.  From here, set the post's publish date and then **Schedule**.

</details>

<details>
 <summary>Flows for saving a published post ⤵️</summary>

* Tap the **Update** button.
* Tap the back button.
* The the three dots to the upper right to get to **Post Settings**.  From here, change the post's publish date and then **Schedule**.</details>

<details>
 <summary>Flows for saving a scheduled post ⤵️</summary>

* Tap the **Schedule** button.
* Tap the back button.
* The the three dots to the upper right to get to **Post Settings**.  From here, change the post's publish date and then **Schedule**.

</details>

### iOS flows

<details>
 <summary>Flows for saving a draft ⤵️</summary>

* Tap the **Update** button.
* Tap the X icon to exit the post and **Update Draft**.
* Tap the three dots to the upper right and then **Publish**.
* The the three dots to the upper right to get to Post Settings. From here, set the post's publish date and then **Schedule**.

</details>

<details>
 <summary>Flows for saving a published post ⤵️</summary>

* Tap the **Update** button.
* Tap the X icon to exit the post and **Update Post**.
* The the three dots to the upper right to get to Post Settings. From here, set the post's publish date and then **Schedule**.

</details>

<details>
 <summary>Flows for saving a scheduled post ⤵️</summary>

* Tap the **Update** button.
* Tap the X icon to exit the post and **Update Post**.
* The the three dots to the upper right to get to Post Settings. From here, set the post's publish date and then **Schedule**.

</details>

SiobhyB committed:
refactor: Remove hack that enabled manual saving

SiobhyB committed:
feat: Detect when post has been saved on native

SiobhyB committed:
docs: Update CHANGELOG

SiobhyB committed:
Merge branch 'trunk' into rnmobile/detect-when-post-is-saved

SiobhyB committed:
Merge branch 'trunk' into rnmobile/detect-when-post-is-saved

SiobhyB committed:
refactor: Stop saving markup

SiobhyB commented on projects/packages/videopress/src/client/block-editor/hooks/use-sync-media/index.ts:
```diff
@@ -169,6 +166,26 @@ export function useSyncMedia(
+
+	isNative &&
+		useEffect( () => {
+			const handlePostSaveEvent = () => {
```
As noted in [the relevant Gutenberg PR](https://github.com/WordPress/gutenberg/pull/49781), we could potentially connect this to [Gutenberg's isSavingPost selector](https://github.com/WordPress/gutenberg/blob/c475dcb3ffee7627a8e1d2f13f364a7a9b195a61/packages/editor/src/store/selectors.js#L677-L692) in the future. For now, I've proposed this solution in the interest of time and simplicity.

fluiddot commented on projects/packages/videopress/src/client/block-editor/hooks/use-sync-media/index.ts:
```diff
@@ -169,6 +166,26 @@ export function useSyncMedia(
+	// We therefore need to listen to a native event emitter to know when a post has just been saved.
+	const [ postHasBeenJustSavedNative, setPostHasBeenJustSavedNative ] = useState( false );
+
+	isNative &&
```
I'd check this condition inside the hook and bail early if it's not the native version. The reason is that is not recommended to call hooks conditionally ([reference](https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level)).

SiobhyB replied:
Thanks for calling this out! Fixed in https://github.com/Automattic/jetpack/pull/30131/commits/e1fe619c53edef53dbccce47834fbc3135e658c6.

SiobhyB commented on projects/packages/videopress/src/client/block-editor/blocks/video/index.native.ts:
These changes are brought in from https://github.com/Automattic/jetpack/pull/30134, in order to avoid a regression that prevented this PR from being built. https://github.com/Automattic/jetpack/pull/30134 should be merged **before** this one.

fluiddot reviewed (commented):
Tested different flows in both platforms and confirmed that VideoPress metadata is saved in the below cases:

**iOS:**
- Published post - Tap `Update` button and then close the editor. ✅ 
- Published post - Close editor and save unsaved changes. ✅ 
- Draft post - Tap `Update` button and then close the editor. ✅ 
- Draft post - Close editor and save unsaved changes. ✅ 
- Draft post - Tap `Three dots button/Publish` and open the post again. ✅ 
- Scheduled post - Tap `Update` button and then close the editor. ✅ 
- Scheduled post - Close the editor and save unsaved changes. ✅ 

**Android:**
- Published post - Tap `Update` button and open the post again. ✅ 
- Published post - Close the editor with unsaved changes. ✅ 
- Published post - Tap `Schedule` button and open the post again. ✅ 
- Draft post - Tap `Three dots button/Save` and open the post again. ✅ 
- Draft post - Close the editor and save unsaved changes. ✅ 
- Draft post - Tap `Publish` and open the post again. ✅ 
- Scheduled post - Tap `Schedule` and open the post again. ✅ 
- Scheduled post - Close the editor and save unsaved changes. ✅ 
- Scheduled post - Tap `Publish Now` and open the post again. ✅ 

However, I noticed an issue when previewing a post or switching to HTML mode. It seems unrelated to these changes, so I wouldn't block the PR, but we'd need to track it.

### Preview a post
1. Having a VideoPress block with a video.
2. Change some of the settings that will be synced (e.g. title).
3. Tap on the three dots button.
4. Tap on `Preview`.
5. Close the preview.
6. Open the block settings and observe that the settings changes were discarded.

### Switch to HTML mode
1. Having a VideoPress block with a video.
2. Change some of the settings that will be synced (e.g. title).
3. Tap on the three dots button.
4. Tap on `HTML Mode`.
5. Switch back to Visual Mode.
6. Open the block settings and observe that the settings changes were discarded.

Apart from the above-unrelated issue, the PR looks great to me. I only added a comment that would be great to tackle before approving it.

SiobhyB committed:
refactor: Move conditional inside hook

Feedback ref: https://github.com/Automattic/jetpack/pull/30131#discussion_r1169823445

SiobhyB commented:
@fluiddot, thank you for catching those issues with preview and toggle mode! I've gone ahead to create an issue in https://github.com/wordpress-mobile/gutenberg-mobile/issues/5677 and have added it to our project board. 🙇‍♀️

SiobhyB committed:
Merge branch 'trunk' into rnmobile/detect-when-post-is-saved

fluiddot commented on projects/packages/videopress/src/client/block-editor/hooks/use-sync-media/index.ts:
```diff
@@ -169,6 +166,29 @@ export function useSyncMedia(
+			setPostHasBeenJustSavedNative( true );
+		};
+
+		const subscription = subscribePostSaveEvent( handlePostSaveEvent );
```
Nitpick: Since `handlePostSaveEvent` is only referenced here, I think we can simplify this with the following code.
```suggestion
		const subscription = subscribePostSaveEvent( () => setPostHasBeenJustSavedNative( true ) );
```

fluiddot commented on projects/packages/videopress/src/client/block-editor/hooks/use-sync-media/index.ts:
```diff
@@ -169,6 +166,29 @@ export function useSyncMedia(
+		return () => {
+			if ( subscription ) {
+				subscription.remove();
+			}
```
Nitpick: We can use the optional chaining operator here to simplify the code.
```suggestion
			subscription?.remove();
```

fluiddot commented on projects/packages/videopress/src/client/block-editor/hooks/use-sync-media/index.ts:
```diff
@@ -299,12 +321,10 @@ export function useSyncMedia(
-
 		debug( '%o Post has been just saved. Syncing...', attributes?.guid );
 
+		setPostHasBeenJustSavedNative( false );
```
This won't produce any side effect in the web version as `postHasBeenJustSavedNative` will remain `false. However, I'm wondering if we should only execute it in the native version as it's only used there, WDYT?

SiobhyB replied:
Happy to go with your preferences, changed in https://github.com/Automattic/jetpack/pull/30131/commits/011d672780ec4feed58d106ddf27c1edfbd2203d.

fluiddot reviewed (commented):
I re-tested some of the cases shared in my last review. However, I'm trying to test the web version but I'm having trouble to have the VideoPress block registered 🤔.

SiobhyB committed:
refactor: Simplify logic within useEffect hook

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

SiobhyB committed:
refactor: Use chaining operator to simplify code

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

SiobhyB committed:
refactor: Use isNative conditional to limit code

SiobhyB committed:
Merge branch 'rnmobile/detect-when-post-is-saved' of github.com:Automattic/jetpack into rnmobile/detect-when-post-is-saved

fluiddot reviewed (changes requested):
As internally discussed (p1681833498706689/1681831906.251289-slack-C03TA48NSUX), we'd need to move this import to a native file because the package is only available in the native version.

SiobhyB committed:
refactor: Add native file for useSyncMedia hook

SiobhyB committed:
refactor: Remove native functionality in web file

SiobhyB committed:
refactor: Switch from .ts to .js

SiobhyB committed:
refactor: Strip hook down to native functionality

SiobhyB committed:
fix: Revert typo preventing attributes updating

SiobhyB committed:
Merge branch 'trunk' into rnmobile/detect-when-post-is-saved

SiobhyB commented:
I've gone ahead to move the native functionality to a new `.native.js` file. Note, I opted to go with `js` file as we're making some changes to the web's types. In addition, although we're not currently using any strings in this file on native, the fact we can't import translations from `.ts` files swayed my decision. It felt that using a `.js` version would be the most straightforward and future-proof approach. Open to different opinions, though.

The removal of native-specific functionality from the main file can be found in https://github.com/Automattic/jetpack/pull/30131/commits/f050d65352cffb62cf1ef2b2824168df43c02a82. For the new native file, it was possible to remove a bunch of (currently) unused functionality, as shown in https://github.com/Automattic/jetpack/pull/30131/commits/321d34361653428162d6867454acb35189a98038.

SiobhyB committed:
refactor: Revert intro of new line

fluiddot reviewed (approved):
LGTM 🎊 !

I tested the web version and worked as expected ✅ . Nevertheless, it would be great to request a second review from web folks to confirm no regressions are introduced.

Regarding the native version, I re-tested some of the cases shared [here](https://github.com/Automattic/jetpack/pull/30131#pullrequestreview-1389792807) in both platforms and worked as expected ✅.

dhasilva reviewed (approved):
No regressions on web version. Nice work!
